src/snowflake/snowpark/_internal/ast/builder.py

```diff
@@ -0,0 +1,176 @@
+#!/usr/bin/env python3
+#
+# Copyright (c) 2012-2025 Snowflake Computing Inc. All rights reserved.
+#
+
+import base64
+import itertools
+import sys
+import uuid
+from collections import namedtuple, deque
+from typing import Optional, List
+
+import snowflake.snowpark._internal.proto.generated.ast_pb2 as proto
+
+from snowflake.snowpark.version import VERSION
+
+# The current AST version number (generated by the DSL).
+CLIENT_AST_VERSION = proto.__Version__.MAX_VERSION
+
+
+SerializedBatch = namedtuple("SerializedBatch", ["request_id", "batch"])
+
+
+class AstBuilder:
+    """
+    A handler for AST statements that enables a Snowpark oject (e.g. Dataframe, Table, etc.) to carry its own AST,
+    and to generate AST requests.
+
+    The core statement types are:
+    - Bind: Creates a new variable and assigns a value to it.
+    - Eval: Evaluates a variable.
+    """
+
+    # Function used to generate request IDs. This is overridden in some tests.
+    __generate_request_id = uuid.uuid4
+
+    # Class variable for generating IDs for bind statements.
+    __id_gen = itertools.count(start=1)
+
+    # NOTE: itertools.count and its __next__ method are only thread-safe in CPython with GIL.
+    # In other implementations (Jython, PyPy), this may not be thread-safe, and once the GIL is removed from CPython, it will not be thread-safe either.
+    # Likely, it would be best to make all of the methods in this class thread-safe with threading.Rlock
+    def __init__(
+        self,
+    ) -> None:
+        """
+        Initializes a new AST batch.
+        """
+        self._bind: proto.Bind = None
+        self.__eval: proto.Eval = None
+        self.__dependencies: List[AstBuilder] = []
+        self.__required_by: List[AstBuilder] = []
+
+    @classmethod
+    def Bind(cls, symbol: Optional[str] = None) -> "AstBuilder":
+        """
+        Factory method to create an AstBuilder instance with a new bind statement.
+
+        Args:
+            symbol: An optional symbol to name the new variable in the bind statement.
+        Returns:
+            An instance of AstBuilder with a bind statement initialized.
+        """
+        b = cls()
+        b._bind = proto.Bind()
+        b._bind.uid = next(AstBuilder.__id_gen)
+        b._bind.symbol.value = symbol if isinstance(symbol, str) else ""
+        return b
+
+    def add_dependency(
+        self, dependency: "AstBuilder", add_inverse: bool = False
+    ) -> None:
+        """
+        Adds a dependency to the current AstBuilder instance.
+        This allows for tracking dependencies between different AST statements.
+
+        Args:
+            dependency: An AstBuilder instance that is a dependency of the current instance.
+        """
+        if dependency not in self.__dependencies:
+            self.__dependencies.append(dependency)
+        if add_inverse:
+            dependency.add_required_by(self, add_inverse=False)
+
+    def add_required_by(
```

Comment:
what is the difference between "depends_on" and "required_by"?

Reply:
"depends_on" corresponds to child statements for this one (dependencies)
"required_by" corresponds to parent statements that use this one

Reply:
ah, I think you want to have a single mechanism for dependence. You can define a single interface, and manage both with it! I.e.,

Have object A depend on B: A.depends_on(B).  Now the question is, do you need directed depdence? I.e.,

A -> B, B->A, A <-> B? are these different cases?

I would argue, we mostly care about a DAG structure here, i.e. A -> B only. This will make a lot of things easier, because the resulting graph can be topologically ordered (which I believe is what we want for the closure).